crates/tinymist-query/src/code_action.rs

```diff
@@ -116,4 +120,32 @@ mod tests {
             })
         });
     }
+
+    fn compute_code_action_context(
+        ctx: &mut LocalContext,
+        request_range: &LspRange,
+    ) -> CodeActionContext {
+        let Warned { output, warnings } = typst::compile::<PagedDocument>(&ctx.world);
+        let errors = output.err().unwrap_or_default();
+        let compiler_diagnostics = warnings.iter().chain(errors.iter());
+
+        // Run the linter for additional diagnostics as well.
+        let diagnostics = DiagWorker::new(ctx)
+            .check()
+            .convert_all(compiler_diagnostics)
+            .into_values()
+            .flatten();
+
+        CodeActionContext {
+            diagnostics: diagnostics
+                .filter(|diag| ranges_overlap(&diag.range, request_range))
+                .collect(),
+            only: None,
+            trigger_kind: None,
+        }
+    }
+
+    fn ranges_overlap(r1: &LspRange, r2: &LspRange) -> bool {
```

Comment:
If there's already an existing implementation of range overlap elsewhere in the project, let me know and I'll switch to it. (I couldn't find anything when I searched, so I just rolled my own.)